src/Error/functions.php

```diff
@@ -0,0 +1,206 @@
+     * @param bool $showFrom If set to true, the method prints from where the function was called.
+     * @return mixed The same $var that was passed
+     * @link https://book.cakephp.org/4/en/development/debugging.html#basic-debugging
+     * @link https://book.cakephp.org/4/en/core-libraries/global-constants-and-functions.html#debug
+     */
+    function debug($var, $showHtml = null, $showFrom = true)
+    {
+        if (!Configure::read('debug')) {
+            return $var;
+        }
+
+        $location = [];
+        if ($showFrom) {
+            $trace = Debugger::trace(['start' => 0, 'depth' => 1, 'format' => 'array']);
+            if (isset($trace[0]['line']) && isset($trace[0]['file'])) {
+                $location = [
+                    'line' => $trace[0]['line'],
+                    'file' => $trace[0]['file'],
+                ];
+            }
+        }
+
+        Debugger::printVar($var, $location, $showHtml);
+
+        return $var;
+    }
+
+    /**
+     * Outputs a stack trace based on the supplied options.
+     *
+     * ### Options
+     *
+     * - `depth` - The number of stack frames to return. Defaults to 999
+     * - `args` - Should arguments for functions be shown? If true, the arguments for each method call
+     *   will be displayed.
+     * - `start` - The stack frame to start generating a trace from. Defaults to 1
+     *
+     * @param array<string, mixed> $options Format for outputting stack trace
+     * @return void
+     */
+    function stackTrace(array $options = []): void
+    {
+        if (!Configure::read('debug')) {
+            return;
+        }
+
+        $options += ['start' => 0];
+        $options['start']++;
+
+        /** @var string $trace */
+        $trace = Debugger::trace($options);
+        echo $trace;
+    }
+
+    /**
+     * Prints out debug information about given variable and dies.
+     *
+     * Only runs if debug mode is enabled.
+     * It will otherwise just continue code execution and ignore this function.
+     *
+     * @param mixed $var Variable to show debug information for.
+     * @param bool|null $showHtml If set to true, the method prints the debug data in a browser-friendly way.
+     * @return void
+     * @link https://book.cakephp.org/4/en/development/debugging.html#basic-debugging
+     */
+    function dd($var, $showHtml = null): void
+    {
+        if (!Configure::read('debug')) {
+            return;
+        }
+
+        $trace = Debugger::trace(['start' => 1, 'depth' => 2, 'format' => 'array']);
+        /** @psalm-suppress PossiblyInvalidArrayOffset */
+        $location = [
+            'line' => $trace[0]['line'],
+            'file' => $trace[0]['file'],
+        ];
+
+        Debugger::printVar($var, $location, $showHtml);
+        die(1);
+    }
+}
+
+namespace {
+
+    use function Cake\Error\dd as cakeDd;
+    use function Cake\Error\debug as cakeDebug;
+    use function Cake\Error\stackTrace as cakeStackTrace;
```

Comment:
Would the idea be to use 5.x to deprecate the global functions and provide rector rules to do renaming on application code?

Reply:
I am undecided on that. For e.g. the I18n functions which are widely used in apps, I would want them to continue being global functions and not namespaced.

Maybe in 5.x we can make the global function definitions configurable, disabled by default and enabled through app config?

Reply:
Maybe in 5.x we can make the global function definitions configurable, disabled by default and enabled through app config?

That could be interesting. I'm having a hard time figuring out how we do that and maintain backwards compatibility though. Userland code currently assumes that functions are autoloaded. Having a `Configure` value implies that we wouldn't know which function files should be loaded until after `Application::bootstrap()` which means short functions couldn't be used in `Application::bootstrap()` or earlier than than.

One potential idea is to have `src/basics.php` become a 'backwards compatible' entrypoint. We could then have users 'upgrade' by removing `basics.php`, and adding global function files for each package. Users could then load only what they want. It won't be as clean as before but it is clear and explicit, and the magic is still there if you use the namespaced functions.

Reply:
Your suggestion could work. So should I put the namespaced and global functions for each package in separate files?

Reply:
I think that's the only way it could work. Its a bit of boilerplate for us to maintain, but so far global functions have been really low maintenance as they don't change often.

Reply:
> One potential idea is to have src/basics.php become a 'backwards compatible' entrypoint.

The problem with using `src/basics.php` is that it won't be loaded when using split packages and we would need to keep the global functions loaded by default for 4.x.

Reply:
> The problem with using src/basics.php is that it won't be loaded when using split packages

Darn forgot about the split packages :cry: